### DISCUSSION TITLE
use case for redis or jwt

### DISCUSSION THREAD

**Discussion by encryptblockr** [abersheeran/asgi-ratelimit#26] [Q&A]
I see that there is option to use redis or jwt

can you please elaborate on the difference between each use case? 
when using the jwt version, what does that offer?

doc is lacking ready to use examples so hope to get some help here

thanks

**Comment by abersheeran** [comment]
First of all, you need to have a JWT, its authentication algorithm and the key for authentication.

And then, use `jwt_auth = create_jwt_auth("key", "algorithm")` to create a authentication function that the parameter for `RateLimitMiddleware`.

When the request comes in, it will automatically extract the JWT from the request header 'authorization' and try to verify it. Then, `RateLimitMiddleware` use the "user" and "group" information carried in the JWT to match the rules and try to limit the flow.

**Reply by encryptblockr** [reply]
i already use jwt for authenticating users, so i am a bit confused if i should use redis or jwt here
i also already have redis in the stack which serves as backend for celery

what i will like to start with is just to setup rate limit on all endpoints for all users and then later have different rate limits for different user roles/permissions

**Reply by encryptblockr** [reply]
mind sharing a "real" example that uses redis and/or jwt that one can follow? 
real example will be great here and make things very easy to follow and use

**Reply by abersheeran** [reply]
I don't have such a "real example", but I think you can try to use it first, and if there is a problem, you can ask them.

**Reply by encryptblockr** [reply]
you still havent answered the main question i keep asking

**what is difference between the redis and jwt 
and what does the jwt offer? compared to the redis?**

**Reply by encryptblockr** [reply]
ended up using https://pypi.org/project/fastapi-limiter/ and it worked like a charm...very fast to implement
you may want to add "real" examples to increase adoption of your package

**Reply by abersheeran** [reply]
Maybe I misunderstood. But redis is the back end storage and JWT is the authentication method. These are two completely different aspects of things.

But anyway, it's good that you solved your problem.
